### ISSUE TITLE
starting from scratch, solr doesn't initialize properly

### ISSUE THREAD

**Issue by zackw** [berkmancenter/tagteam#37]
Again, I'm trying to initialize my own instance of tagteam, following the instructions in the README.  I have gotten to the point where loading https://[my-instance]/ runs the application.  It then appears to make an internal query to Solr, which fails, because Solr isn't properly configured, I think?  I get this error in the Rails `production.log`

```
A ActionView::Template::Error occurred in hubs#home:

  RSolr::Error::Http - 404 Not Found
Error:     Not Found

URI: http://localhost:8982/solr/production/select?wt=json
Request Headers: {"Content-Type"=>"application/x-www-form-urlencoded; charset=UTF-8"}
Request Data: "fq=type%3AFeedItem&start=0&rows=0&facet=true&f.hub_ids_im.facet.limit=4&f.hub_ids_im.facet.mincount=1&facet.field=hub_ids_im&q=*%3A*"

Backtrace: /home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/rsolr-2.0.2/lib/rsolr/client.rb:195:in `rescue in execute'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/rsolr-2.0.2/lib/rsolr/client.rb:185:in `execute'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/rsolr-2.0.2/lib/rsolr/client.rb:180:in `send_and_receive'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/sunspot_rails-2.2.7/lib/sunspot/rails/solr_instrumentation.rb:16:in `block in send_and_receive_with_as_instrumentation'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.3/lib/active_support/notifications.rb:164:in `block in instrument'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.3/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.3/lib/active_support/notifications.rb:164:in `instrument'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/sunspot_rails-2.2.7/lib/sunspot/rails/solr_instrumentation.rb:15:in `send_and_receive_with_as_instrumentation'
(eval):2:in `post'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/sunspot-2.2.7/lib/sunspot/search/abstract_search.rb:49:in `execute'
/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/sunspot_rails-2.2.7/lib/sunspot/rails/searchable.rb:389:in `solr_execute_search'
  app/models/hub.rb:143:in `most_active_hubs'
```

There is also a `sunspot-solr-production.log` in the app logs directory, containing two copies of this error:

```
Oct 25, 2017 4:14:12 PM org.apache.solr.core.CoreContainer recordAndThrow
SEVERE: Unable to create core: collection1
org.apache.solr.common.SolrException: Could not load config for solrconfig.xml
        at org.apache.solr.core.CoreContainer.createFromLocal(CoreContainer.java:991)
        at org.apache.solr.core.CoreContainer.create(CoreContainer.java:1051)
        at org.apache.solr.core.CoreContainer$3.call(CoreContainer.java:634)
        at org.apache.solr.core.CoreContainer$3.call(CoreContainer.java:629)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Can't find resource 'solrconfig.xml' in classpath or '/home/tagteam/tagteam/solr/collection1/conf/', cwd=/home/tagteam/tagteam/vendor/bundle/ruby/2.3.0/gems/sunspot_solr-2.2.0/solr
        at org.apache.solr.core.SolrResourceLoader.openResource(SolrResourceLoader.java:318)
        at org.apache.solr.core.SolrResourceLoader.openConfig(SolrResourceLoader.java:283)
        at org.apache.solr.core.Config.<init>(Config.java:103)
        at org.apache.solr.core.Config.<init>(Config.java:73)
        at org.apache.solr.core.SolrConfig.<init>(SolrConfig.java:117)
        at org.apache.solr.core.CoreContainer.createFromLocal(CoreContainer.java:989)
        ... 9 more
```

I observed that there is no 'collection1' directory below .../tagteam/solr, so I tried creating it and copying tagteam/solr/conf into that directory; that prevents any errors being written to  `sunspot-solr-production.log`, but a manual GET to  `http://localhost:8982/solr/production/select?wt=json` still returns a 404 error, and if I instead GET `http://localhost:8982/solr/`, it returns an HTML page containing an error message:

```
      <div id="init-failures">
          <h2>SolrCore Initialization Failures</h2>
          <ul></ul>
          <p>Please check your logs for more information</p>
      </div>
```

As there is now nothing in the logs, I am stuck at this point.

**Comment by jsdiaz** [comment]
I think you can fix this by navigating to the folder where your solr binary files are (or if you installed it from a pkg mgr, without the 'bin/' from any directory) and entering the following command: bin/solr create_core -c name_of_core_or_collection

**Comment by zackw** [comment]
Please understand that this is the first time I have ever worked with Rails or any of the other software involved.  I understand what you are telling me to do except for the `name_of_core_or_collection` part.  Is that literally what I should type?  If not, what should I type?  `collection1` maybe?

**Comment by zackw** [comment]
I attempted to fill in the gaps in the above instructions and got hopelessly stuck.

* There is no executable called `solr` anywhere in my installation (which contains only the things that the README says are necessary, nothing else).  There is an executable called `sunspot-solr` but it does not accept a `-c` option.
* I understand that Sunspot is a Ruby client library for Solr.  I looked up its usage instructions but didn't find any advice for this problem; it makes it sound like this step shouldn't be needed.
* I managed to create a "core", whatever that is, using the HTML administrative interface exposed by Solr once it's up and running. This created a `collection1` directory inside `tagteam/solr` with some index files inside.  However, upon shutting down and restarting Solr I still get the "cannot create a core" errors, and the "SolrCore initialization failure" message from `http://localhost:8982/solr`, and the 404 from `http://localhost:8982/solr/production/select?wt=json`.

**Comment by peter-hank** [comment]
@zackw have you run `rake sunspot:solr:start`?

**Comment by zackw** [comment]
@peter-hank Yes, that command is what causes `sunspot-solr-production.log` to be created with the "SEVERE: Unable to create core: collection1" backtrace I showed up above.

**Comment by zackw** [comment]
I have been tinkering with this for the past two hours and I now have a better understanding of what is wrong, but I still don't understand how to fix it properly.  (Again, please keep in mind that I have no prior experience with any of these technologies.  This is all my best guess based on the scanty internet documentation and a lot of experimentation.)

 * In production mode, the Rails application expects to be able to query Solr at URLs of the form `http://localhost:8982/solr/production/select` with various query parameters.
 * From Solr's perspective, for that to work, there must exist a "core" (this seems to be Solr's name for a set of indexes, like how a "database" is a set of tables in SQL-ese) named "production".
 * `sunspot.yml` _intends_ to set that up with its `path: /solr/production` setting, but as far as I can tell, that setting has no effect whatsoever.
 * Sunspot, somehow (I cannot find the code that does this) is directing Solr that on startup it should initialize a core named `collection1`, *instead of* `production` or anything like that.  (Possibly this is just Solr's default?)
 * Solr refuses to actually create the necessary files and directories for `collection1`; if they don't already exist, it bombs out.
 * In particular, the files checked into Git as `tagteam/solr/conf/*` must exist as `tagteam/solr/collection1/conf/*` for Solr to initialize successfully.
 * An unrelated but equally serious issue is that Sunspot insists on telling Solr that its _data_ directory should be `/tmp`, which, in my installation, does not have anything like the space that will be required for that.  There is code inside Sunspot that can be used to override this (set the `solr_data_dir` property on the Sunspot "Server" object) but I was not able to find any way to set that property from `sunspot.yml`.
 * `rake sunspot:solr:start` fails to detect any of the above problems; it says "Solr started successfully" no matter what.

I know how to kludge around _some_ of these problems:

 * Create a symbolic link `tagteam/solr/collection1` pointing to `tagteam/solr`.
 * Edit `tagteam/solr/conf/solrconfig.xml` and change the `<dataDir>` setting to point to a more appropriate location.
 * Now `RAILS_ENV=production rake sunspot:solr:start` will activate a Solr instance that is actually fully initialized and functional.  However, it still won't respond to queries at the URLs that Tagteam wants to use.

What I still don't know how to do is tell Sunspot and/or Solr that that initial core should be named 'production', not 'collection1'.

**Comment by jsdiaz** [comment]
Zack, I would suggest you reinstall solr following something like this tutorial depending on which platform you are running it on:

https://www.howtoforge.com/tutorial/how-to-install-and-configure-solr-on-ubuntu-1604/

Once you have solr running on its own, prior to getting TT pointed to it, then I think we can proceed to help you with TT related issues.

**Comment by zackw** [comment]
@jsdiaz I don't see how that would help, since Sunspot embeds its own copy of Solr and will insist on using it, with the messed-up configuration, won't it?

**Comment by zackw** [comment]
I have it working now!  The critical missing piece was a `solr.xml` in `tagteam/solr` listing the expected cores, along the lines of what is described in https://github.com/sunspot/sunspot/issues/115 .

**Comment by patricklewis** [comment]
Fixed in dev branch with 04183a1.
